### PR TITLE
Split IMediator interface in IPublisher and ISender interface

### PR DESCRIPTION
**Pull request by KenBonny**
The reason for this change is the quick [twitter chat](https://twitter.com/bonny_ken/status/1309182617653305344) I had with Jimmy about this.

### CODE REVIEW

**Comment by KenBonny** [comment]
Oh yes, I've removed unused usings as this is a pet peeve of mine. 😃

File: test/MediatR.Tests/PipelineTests.cs
Comment (lilasquared) [inline test/MediatR.Tests/PipelineTests.cs:11]: I understand removing unused references, but I don't know the implications of moving usings from inside the namespace to outside, any reason for doing so?

**Comment by lilasquared** [comment]
@KenBonny once this is merged the dependency injection package should be updated as well to register the publisher and sender. https://github.com/jbogard/MediatR.Extensions.Microsoft.DependencyInjection

File: test/MediatR.Tests/PipelineTests.cs
Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs:11]: There is no benefit to using a `using` statement inside or outside of the namespace. I think the only difference is the scope. So the `using`s inside the namespace are only accessible inside the namespace and the `using`s outside are for the whole file. Since there is only one namespace in this file, it won't matter.

When I saw this in the file, my neurotic nature wanted to put them together as they belong together.

Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs:11]: If you want, I can go over the code and put them all together and order them alphabetically.

Comment (lilasquared) [reply test/MediatR.Tests/PipelineTests.cs:11]: If they are functionally the same it doesn't matter to me, I'll defer to Jimmy.  I've always put them outside but have seen them inside namespaces as well.

**Comment by KenBonny** [comment]
@lilasquared Thanks for the heads up. I think the only thing that needs to happen is for the MediatR package to be updated. If the `IMediator` interface is registered, then I do believe most functionality should work. I did see that there is a line that registers all types in an assembly instead of registering everything by hand. I think that takes care of registering every type and interface in the assembly.

File: test/MediatR.Tests/PipelineTests.cs
Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs:11]: Ok, I'll leave it up to @jbogard to decide if I roll this back, if I format and order the rest of the `using`s in the codebase or if this is ok.

**Comment by lilasquared** [comment]
I'm pretty sure that only scans and searches for implementations of the handlers.  I'm referring to https://github.com/jbogard/MediatR.Extensions.Microsoft.DependencyInjection/blob/master/src/MediatR.Extensions.Microsoft.DependencyInjection/Registration/ServiceRegistrar.cs#L220
where the `IMediator` type is registered explicitly.  If you test your branch with that package and you try to inject `ISender` it fails to resolve.

File: test/MediatR.Tests/PipelineTests.cs
Comment (jbogard) [reply test/MediatR.Tests/PipelineTests.cs:11]: I prefer consistency tbqh

Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs:11]: Would you like me to roll it back or would you like me to look through the codebase and format the other places?

Comment (jbogard) [reply test/MediatR.Tests/PipelineTests.cs:11]: If it's not consistent in the codebase today, leave it alone.

Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs]: Should be rolled back now.

File: test/MediatR.Tests/PipelineTests.cs
Comment (no1melman) [inline test/MediatR.Tests/PipelineTests.cs:2]: I kinda feel that this whole file should be reverted, bar the removing of the namespace - just inconsistent formatting

Comment (KenBonny) [reply test/MediatR.Tests/PipelineTests.cs:2]: I rolled it back. On save, my IDE (Rider) formats all open files to my coding standards.

Comment (jbogard) [reply test/MediatR.Tests/PipelineTests.cs:2]: So basically I need an `.editorconfig`

**Review by jbogard — APPROVED** [review]

**Comment by KenBonny** [comment]
Is this ok, can it be merged or do other steps need to happen? That the other reviewers approve, for example. Or do I need to press a button to merge the PR?

**Comment by KenBonny** [comment]
Don't mistake my question for impatience, but when will this be available in the package?

I'm asking so I can also update the DI package @lilasquared mentioned. I have a leaky memory (I should get that looked after) and if this takes a month or 2 (which is fine 😄 ), then I need to add a calendar notification to remind me to do the DI update then.

**Comment by jbogard** [comment]
Tagged! Should be on NuGet shortly.

**Comment by davidroth** [comment]
@jbogard  If you make such changes, please consider increasing the major version instead of the minor version in the future  ;-)

This interface split is technically a breaking change.
For example if an assembly "A" references IMediator  from 8.0.2, but at runtime 8.2. is supplied (assembly unification at build time), the interface split leads to the following runtime exception in "A" if it has not been recompiled with 8.2. 

```csharp
System.MissingMethodException : Method not found: 'System.Threading.Tasks.Task`1<!!0> MediatR.IMediator.Send(MediatR.IRequest`1<!!0>, System.Threading.CancellationToken)'.
```

Thanks 😃

**Pull request merged**
